File: src/main/java/pl/training/workshop/m6/s12_onemany/step2/CancellationDesk.java
package pl.training.workshop.m6.s12_onemany.step2;

import java.time.LocalDateTime;
import java.util.List;

import pl.training.workshop.m6.s12_onemany.TicketData;
import pl.training.workshop.shared.Money;

/** Krok 2: klient przeniesiony na nowy kontrakt; rozróżnienie jeden/wiele jeszcze widać. */
public final class CancellationDesk {
    private final RefundService service = new RefundService();

    public Money refund(List<TicketData> tickets, LocalDateTime now) {
        Refundable refundable = tickets.size() == 1
                ? new SingleTicket(tickets.getFirst())
                : TicketGroup.of(tickets);
        return service.refund(refundable, now);
    }
}

File: typescript/src/workshop/m6/s12_onemany/step2/CancellationDesk.ts
import type { Money } from '../../../shared/Money.js';
import type { LocalDateTime } from '../../../shared/time.js';
import type { TicketData } from '../TicketData.js';
import type { Refundable } from './Refundable.js';
import { RefundService } from './RefundService.js';
import { SingleTicket } from './SingleTicket.js';
import { TicketGroup } from './TicketGroup.js';

/** Krok 2: klient przeniesiony na nowy kontrakt; rozróżnienie jeden/wiele jeszcze widać. */
export class CancellationDesk {
  private readonly service = new RefundService();

  refund(tickets: readonly TicketData[], now: LocalDateTime): Money {
    const [first] = tickets;
    const refundable: Refundable = tickets.length === 1 && first !== undefined
      ? new SingleTicket(first)
      : TicketGroup.of(tickets);
    return this.service.refund(refundable, now);
  }
}
